### PR TITLE
Fix Nemotron W4A16 NVFP4 MoE backend

### PR DESCRIPTION
**Pull request by danielafrimi**
Detect Nemotron-H ModelOpt mixed-precision checkpoints with W4A16_NVFP4 MoE expert layers.

- Keep the existing SM100 `flashinfer_trtllm` default for regular NVFP4/W4A4 MoE, but route W4A16 NVFP4 expert MoE to `Marlin`. 
- Add focused override tests for both W4A16 NVFP4 and regular NVFP4 paths.

Also support TP-sharded W4A16 NVFP4 Marlin layers whose local K/N dimensions do not meet a Marlin thread-tile alignment. Pad weights, scales, and activations to the smallest valid Marlin tile shape.

This mirrors vLLM’s behavior: vLLM keeps W4A16 NVFP4 as a distinct ModelOpt path and calls NVFP4 MoE backend selection with activation_key=None. That filters out W4A4-style FlashInfer/TRTLLM backends and leaves the Marlin W4A16 path. T

---
### CI States

Latest PR Test (Base): :x: [Run #31096262776](https://github.com/sgl-project/sglang/actions/runs/31096262776)
Latest PR Test (Extra): :x: [Run #31096262312](https://github.com/sgl-project/sglang/actions/runs/31096262312)

### CODE REVIEW

File: python/sglang/srt/arg_groups/overrides.py [hunk 1]
```diff
@@ -939,6 +939,25 @@ def _inkling_overrides(server_args: Any, hf_config: Any) -> dict:
     return overrides
 
 
+def _has_modelopt_w4a16_moe_layers(hf_config: Any) -> bool:
```
Comment (mmangkad) [inline python/sglang/srt/arg_groups/overrides.py]: Inline and remove defensive attr checks

File: python/sglang/srt/arg_groups/overrides.py [hunk 2]
```diff
@@ -973,10 +992,17 @@ def _nemotron_h_overrides(server_args: Any, hf_config: Any) -> dict:
         server_args.moe_runner_backend == "auto"
     ):
         if is_sm100_supported() and server_args.moe_a2a_backend == "none":
```
Comment (mmangkad) [inline python/sglang/srt/arg_groups/overrides.py:1263]: Handle w4a16 before the a2a branch and reject unsupported combinations

File: python/sglang/srt/layers/quantization/marlin_utils_fp4.py [hunk 3]
```diff
@@ -138,11 +150,40 @@ def prepare_nvfp4_layer_for_marlin(layer: torch.nn.Module) -> None:
...
+            torch.nn.functional.pad(layer.weight, (0, pad_cols, 0, pad_rows)),
+            requires_grad=False,
         )
+        layer.weight_scale = torch.nn.Parameter(
+            torch.nn.functional.pad(
+                layer.weight_scale, (0, scale_pad_cols, 0, pad_rows)
+            ),
+            requires_grad=False,
+        )
+        if hasattr(layer, "bias") and layer.bias is not None:
```
Comment (mmangkad) [inline python/sglang/srt/layers/quantization/marlin_utils_fp4.py]: Can be consolidated with existing bias handling

File: python/sglang/srt/layers/quantization/marlin_utils_fp4.py [hunk 4]
```diff
@@ -138,11 +150,40 @@ def prepare_nvfp4_layer_for_marlin(layer: torch.nn.Module) -> None:
...
+    padded_size_n, padded_size_k = min(
+        (
+            ((part_size_n + 63) // 64 * 64, (part_size_k + 127) // 128 * 128),
+            ((part_size_n + 127) // 128 * 128, (part_size_k + 63) // 64 * 64),
+        ),
+        key=lambda nk: (nk[0] * nk[1], nk[0] + nk[1]),
+    )
+
+    if (padded_size_n, padded_size_k) != (part_size_n, part_size_k):
+        logger.warning_once(
```
Comment (mmangkad) [inline python/sglang/srt/layers/quantization/marlin_utils_fp4.py]: I think warnings should generally be actionable, and since this is expected behavior that adds startup noise, maybe we can remove it

**Comment by b8zhong** [comment]
/rerun-test test/registered/4-gpu-models/test_nvidia_nemotron_3_super_nvfp4.py

File: python/sglang/srt/arg_groups/overrides.py (hunk 2 above)
Comment (danielafrimi) [reply python/sglang/srt/arg_groups/overrides.py:1263]: done

File: python/sglang/srt/arg_groups/overrides.py (hunk 1 above)
Comment (danielafrimi) [reply python/sglang/srt/arg_groups/overrides.py]: fixed

File: python/sglang/srt/layers/quantization/marlin_utils_fp4.py (hunk 3 above)
Comment (danielafrimi) [reply python/sglang/srt/layers/quantization/marlin_utils_fp4.py]: done

File: python/sglang/srt/layers/quantization/marlin_utils_fp4.py (hunk 4 above)
Comment (danielafrimi) [reply python/sglang/srt/layers/quantization/marlin_utils_fp4.py]: removed this

**Comment by danielafrimi** [comment]
@b8zhong thanks for the review - fixed all

**Comment by mmangkad** [comment]
@danielafrimi can you fix the lint failure?

**Review by b8zhong — APPROVED** [review]

**Comment by mmangkad** [comment]
/rerun-test test/registered/models_e2e/test_nvidia_nemotron_3_nano.py

File: python/sglang/srt/arg_groups/overrides.py [hunk 5]
```diff
@@ -1186,6 +1186,12 @@ def _nemotron_h_overrides(server_args: Any, hf_config: Any) -> dict:
         "modelopt_mixed",
     ]
     quantization = server_args.quantization
+    has_w4a16_moe_layers = is_modelopt and any(
+        info.get("quant_algo") == "W4A16_NVFP4" and ".experts." in name
+        for name, info in hf_config.quantization_config.get(
+            "quantized_layers", {}
+        ).items()
+    )
```
Comment (mmangkad) [inline python/sglang/srt/arg_groups/overrides.py]: Move this below quantization normalization and only run it for `modelopt_mixed` to avoid breaking fp8/nvfp4 configs without an embedded `quantization_config`

**Review by mmangkad — CHANGES_REQUESTED** [review]

File: python/sglang/srt/arg_groups/overrides.py (hunk 5 above)
Comment (danielafrimi) [reply python/sglang/srt/arg_groups/overrides.py]: done

**Review by mmangkad — APPROVED** [review]
